aten/src/ATen/templates/TensorBody.h

```diff
@@ -358,12 +358,26 @@ class CAFFE2_API Tensor {
   /// TensorOptions.h.
   TensorOptions options() const;
 
-  void* data_ptr() const {
+  void* data_ptr() const & {
     return this->unsafeGetTensorImpl()->data();
   }
 
+  void* data_ptr() const && = delete;
+
+  template <typename T>
+  T * data_ptr() const &;
+
   template <typename T>
-  T * data_ptr() const;
+  T * data_ptr() const && {
+    static_assert(
+        !std::is_same<T, T>::value,
+        "Calling data_ptr() on temporary is potentially unsafe and can "
+        "indicate use-after-free bug if it was the last reference to the tensor"
+        " (as in `contiguous().data_ptr()`). Please store result in a variable "
```

Comment:
"Please store result in variable first" -> "Please store the Tensor result in a variable before accessing the data pointer, e.g. "Tensor t_contig = t.contiguous(); t_contig.data_ptr()`